Hacks to reduce code size, mostly by eliminating strings.
Currently saves 4KB of code size.

Index: elua-svn/SConstruct
===================================================================
--- elua-svn.orig/SConstruct	2010-12-14 18:24:03.873245001 +0100
+++ elua-svn/SConstruct	2010-12-14 18:24:12.797245001 +0100
@@ -405,6 +405,8 @@
 
   # Optimizer flags (speed or size)
   comp.Append(CCFLAGS = ['-Os','-fomit-frame-pointer'])
+  # Code size reduction flags
+  comp.Append(CCFLAGS = ['-DLUAC_TRUST_BINARIES'])
   #opt += " -ffreestanding"
   #opt += " -fconserve-stack" # conserve stack at potential speed cost, >=GCC4.4
 
Index: elua-svn/src/lua/lauxlib.c
===================================================================
--- elua-svn.orig/src/lua/lauxlib.c	2010-12-14 18:24:03.989245001 +0100
+++ elua-svn/src/lua/lauxlib.c	2010-12-14 18:24:12.797245001 +0100
@@ -158,7 +158,7 @@
 
 LUALIB_API void luaL_checkstack (lua_State *L, int space, const char *mes) {
   if (!lua_checkstack(L, space))
-    luaL_error(L, "stack overflow (%s)", mes);
+    luaL_error(L, "stack overflow");
 }
 
 
Index: elua-svn/src/lua/lbaselib.c
===================================================================
--- elua-svn.orig/src/lua/lbaselib.c	2010-12-14 18:24:03.969245001 +0100
+++ elua-svn/src/lua/lbaselib.c	2010-12-14 18:24:12.801245001 +0100
@@ -349,7 +349,7 @@
   if (i > e) return 0;  /* empty range */
   n = e - i + 1;  /* number of elements */
   if (n <= 0 || !lua_checkstack(L, n))  /* n <= 0 means arith. overflow */
-    return luaL_error(L, "too many results to unpack");
+    return luaL_error(L, "too many results");
   lua_rawgeti(L, 1, i);  /* push arg[i] (avoiding overflow problems) */
   while (i++ < e)  /* push arg[i + 1...e] */
     lua_rawgeti(L, 1, i);
@@ -558,7 +558,7 @@
 static int auxresume (lua_State *L, lua_State *co, int narg) {
   int status = costatus(L, co);
   if (!lua_checkstack(co, narg))
-    luaL_error(L, "too many arguments to resume");
+    luaL_error(L, "too many arguments");
   if (status != CO_SUS) {
     lua_pushfstring(L, "cannot resume %s coroutine", statnames[status]);
     return -1;  /* error flag */
@@ -569,7 +569,7 @@
   if (status == 0 || status == LUA_YIELD) {
     int nres = lua_gettop(co);
     if (!lua_checkstack(L, nres + 1))
-      luaL_error(L, "too many results to resume");
+      luaL_error(L, "too many results");
     lua_xmove(co, L, nres);  /* move yielded values */
     return nres;
   }
Index: elua-svn/src/lua/ldo.c
===================================================================
--- elua-svn.orig/src/lua/ldo.c	2010-12-14 18:24:04.009245001 +0100
+++ elua-svn/src/lua/ldo.c	2010-12-14 18:29:24.445245001 +0100
@@ -386,7 +386,7 @@
 void luaD_call (lua_State *L, StkId func, int nResults) {
   if (++L->nCcalls >= LUAI_MAXCCALLS) {
     if (L->nCcalls == LUAI_MAXCCALLS)
-      luaG_runerror(L, "C stack overflow");
+      luaG_runerror(L, "stack overflow");
     else if (L->nCcalls >= (LUAI_MAXCCALLS + (LUAI_MAXCCALLS>>3)))
       luaD_throw(L, LUA_ERRERR);  /* error while handing stack error */
   }
@@ -437,7 +437,7 @@
   if (L->status != LUA_YIELD && (L->status != 0 || L->ci != L->base_ci))
       return resume_error(L, "cannot resume non-suspended coroutine");
   if (L->nCcalls >= LUAI_MAXCCALLS)
-    return resume_error(L, "C stack overflow");
+    return resume_error(L, "stack overflow");
   luai_userstateresume(L, nargs);
   lua_assert(L->errfunc == 0);
   L->baseCcalls = ++L->nCcalls;
Index: elua-svn/src/lua/loadlib.c
===================================================================
--- elua-svn.orig/src/lua/loadlib.c	2010-12-14 18:24:04.053245001 +0100
+++ elua-svn/src/lua/loadlib.c	2010-12-14 18:24:12.801245001 +0100
@@ -436,7 +436,7 @@
   const char *name = luaL_checkstring(L, 1);
   lua_getfield(L, LUA_ENVIRONINDEX, "preload");
   if (!lua_istable(L, -1))
-    luaL_error(L, LUA_QL("package.preload") " must be a table");
+    luaL_error(L, "%s must be a table", LUA_QL("package.preload"));
   lua_getfield(L, -1, name);
   if (lua_isnil(L, -1))  /* not found? */
     lua_pushfstring(L, "\n\tno field package.preload['%s']", name);
@@ -468,7 +468,7 @@
   /* else must load it; iterate over available loaders */
   lua_getfield(L, LUA_ENVIRONINDEX, "loaders");
   if (!lua_istable(L, -1))
-    luaL_error(L, LUA_QL("package.loaders") " must be a table");
+    luaL_error(L, "%s must be a table", LUA_QL("package.loaders"));
   lua_pushliteral(L, "");  /* error message accumulator */
   for (i=1; ; i++) {
     lua_rawgeti(L, -2, i);  /* get a loader */
Index: elua-svn/src/lua/lstrlib.c
===================================================================
--- elua-svn.orig/src/lua/lstrlib.c	2010-12-14 18:24:04.081245001 +0100
+++ elua-svn/src/lua/lstrlib.c	2010-12-14 18:29:24.065245001 +0100
@@ -203,14 +203,14 @@
   switch (*p++) {
     case L_ESC: {
       if (*p == '\0')
-        luaL_error(ms->L, "malformed pattern (ends with " LUA_QL("%%") ")");
+        luaL_error(ms->L, "malformed pattern");
       return p+1;
     }
     case '[': {
       if (*p == '^') p++;
       do {  /* look for a `]' */
         if (*p == '\0')
-          luaL_error(ms->L, "malformed pattern (missing " LUA_QL("]") ")");
+          luaL_error(ms->L, "malformed pattern");
         if (*(p++) == L_ESC && *p != '\0')
           p++;  /* skip escapes (e.g. `%]') */
       } while (*p != ']');
@@ -637,7 +637,7 @@
     lua_pushlstring(L, s, e - s);  /* keep original text */
   }
   else if (!lua_isstring(L, -1))
-    luaL_error(L, "invalid replacement value (a %s)", luaL_typename(L, -1)); 
+    luaL_error(L, "invalid value");
   luaL_addvalue(b);  /* add result to accumulator */
 }
 
@@ -729,7 +729,7 @@
   const char *p = strfrmt;
   while (*p != '\0' && strchr(FLAGS, *p) != NULL) p++;  /* skip flags */
   if ((size_t)(p - strfrmt) >= sizeof(FLAGS))
-    luaL_error(L, "invalid format (repeated flags)");
+    luaL_error(L, "invalid format");
   if (isdigit(uchar(*p))) p++;  /* skip width */
   if (isdigit(uchar(*p))) p++;  /* (2 digits at most) */
   if (*p == '.') {
@@ -738,7 +738,7 @@
     if (isdigit(uchar(*p))) p++;  /* (2 digits at most) */
   }
   if (isdigit(uchar(*p)))
-    luaL_error(L, "invalid format (width or precision too long)");
+    luaL_error(L, "invalid format");
   *(form++) = '%';
   strncpy(form, strfrmt, p - strfrmt + 1);
   form += p - strfrmt + 1;
@@ -815,8 +815,7 @@
           }
         }
         default: {  /* also treat cases `pnLlh' */
-          return luaL_error(L, "invalid option " LUA_QL("%%%c") " to "
-                               LUA_QL("format"), *(strfrmt - 1));
+          return luaL_error(L, "invalid option");
         }
       }
       luaL_addlstring(&b, buff, strlen(buff));
Index: elua-svn/src/lua/ltablib.c
===================================================================
--- elua-svn.orig/src/lua/ltablib.c	2010-12-14 18:24:03.945245001 +0100
+++ elua-svn/src/lua/ltablib.c	2010-12-14 18:24:12.805245001 +0100
@@ -136,8 +136,7 @@
 static void addfield (lua_State *L, luaL_Buffer *b, int i) {
   lua_rawgeti(L, 1, i);
   if (!lua_isstring(L, -1))
-    luaL_error(L, "invalid value (%s) at index %d in table for "
-                  LUA_QL("concat"), luaL_typename(L, -1), i);
+    luaL_error(L, "invalid value");
     luaL_addvalue(b);
 }
 
Index: elua-svn/src/lua/lua.c
===================================================================
--- elua-svn.orig/src/lua/lua.c	2010-12-14 18:24:04.029245001 +0100
+++ elua-svn/src/lua/lua.c	2010-12-14 18:29:24.509245001 +0100
@@ -40,18 +40,7 @@
 
 
 static void print_usage (void) {
-  fprintf(stderr,
-  "usage: %s [options] [script [args]].\n"
-  "Available options are:\n"
-  "  -e stat  execute string " LUA_QL("stat") "\n"
-  "  -l name  require library " LUA_QL("name") "\n"
-  "  -m limit set memory limit. (units are in Kbytes)\n"
-  "  -i       enter interactive mode after executing " LUA_QL("script") "\n"
-  "  -v       show version information\n"
-  "  --       stop handling options\n"
-  "  -        execute stdin and stop handling options\n"
-  ,
-  progname);
+  fprintf(stderr, "usage: %s [options] [script [args]].\n", progname);
   fflush(stderr);
 }
 
@@ -120,7 +109,7 @@
   int argc = 0;
   while (argv[argc]) argc++;  /* count total number of arguments */
   narg = argc - (n + 1);  /* number of arguments to the script */
-  luaL_checkstack(L, narg + 3, "too many arguments to script");
+  luaL_checkstack(L, narg + 3, "too many arguments");
   for (i=n+1; i < argc; i++)
     lua_pushstring(L, argv[i]);
   lua_createtable(L, narg, n + 1);
Index: elua-svn/src/platform/avr32/MIZAR32/mizar32_conf.h
===================================================================
--- elua-svn.orig/src/platform/avr32/MIZAR32/mizar32_conf.h	2010-12-14 18:24:03.901245001 +0100
+++ elua-svn/src/platform/avr32/MIZAR32/mizar32_conf.h	2010-12-14 18:29:24.725245001 +0100
@@ -143,4 +143,14 @@
 // *****************************************************************************
 // CPU constants that should be exposed to the eLua "cpu" module
 
+// *****************************************************************************
+// Save 3Kb of space in the binary by not loading the errno string table
+#include <stdio.h> /* for sprintf() */
+char *strerror(int e)
+{
+	static char buf[10];
+	(void)sprintf(buf, "errno=%d", e);
+	return(buf);
+}
+
 #endif // #ifndef __MIZAR32_CONF_H__
Index: elua-svn/src/platform/avr32/conf.py
===================================================================
--- elua-svn.orig/src/platform/avr32/conf.py	2010-12-14 18:24:03.925245001 +0100
+++ elua-svn/src/platform/avr32/conf.py	2010-12-14 18:24:12.809245001 +0100
@@ -21,7 +21,7 @@
 ldscript = "src/platform/%s/%s.ld" % ( platform, comp[ 'cpu' ].lower() )
 
 # Standard GCC Flags
-comp.Append(CCFLAGS = ['-ffunction-sections','-fdata-sections','-fno-strict-aliasing','-Wall'])
+comp.Append(CCFLAGS = ['-ffunction-sections','-fdata-sections','-fno-strict-aliasing','-mrelax','-fno-common','-Wall'])
 comp.Append(LINKFLAGS = ['-nostartfiles','-nostdlib','-Wl,--gc-sections','-Wl,--allow-multiple-definition','-Wl,--relax','-Wl,--direct-data','-T',ldscript])
 comp.Append(ASFLAGS = ['-x','assembler-with-cpp','-c'])
 comp.Append(LIBS = ['c','gcc','m'])
Index: elua-svn/src/shell.c
===================================================================
--- elua-svn.orig/src/shell.c	2010-12-14 18:25:02.693245000 +0100
+++ elua-svn/src/shell.c	2010-12-14 18:28:48.037245001 +0100
@@ -48,13 +48,15 @@
 {
   args = args;
   printf( "Shell commands:\n" );
-  printf( "  exit        - exit from this shell\n" );
-  printf( "  help        - print this help\n" );
-  printf( "  ls or dir   - lists filesystems files and sizes\n" );
-  printf( "  cat or type - lists file contents\n" );
-  printf( "  lua [args]  - run Lua with the given arguments\n" );
-  printf( "  recv        - receive a file via XMODEM and execute it\n" );
-  printf( "  ver         - print eLua version\n" );
+  printf( "  exit       - exit from this shell\n" );
+  printf( "  help       - print this help\n" );
+  printf( "  ls         - lists filesystems files and sizes\n" );
+  printf( "  cat        - lists file contents\n" );
+  printf( "  lua [args] - run Lua with the given arguments\n" );
+#ifdef BUILD_XMODEM
+  printf( "  recv       - receive a file via XMODEM and execute it\n" );
+#endif
+  printf( "  ver        - print eLua version\n" );
 }
 
 // 'lua' handler
@@ -103,16 +105,12 @@
   clearerr( stdin );
 }
 
+#ifdef BUILD_XMODEM
 // 'recv' handler
 static void shell_recv( char* args )
 {
   args = args;
 
-#ifndef BUILD_XMODEM
-  printf( "XMODEM support not compiled, unable to recv\n" );
-  return;
-#else // #ifndef BUILD_XMODEM
-
   char *p;
   long actsize;
   lua_State* L;
@@ -157,8 +155,8 @@
   lua_close( L );
   free( shell_prog );
   shell_prog = NULL;
-#endif // #ifndef BUILD_XMODEM
 }
+#endif // #ifdef BUILD_XMODEM
 
 // 'ver' handler
 static void shell_ver( char* args )
@@ -168,7 +166,7 @@
   printf( "For more information visit www.eluaproject.net and wiki.eluaproject.net\n" );
 }
 
-// 'ls' and 'dir' handler
+// 'ls' handler
 static void shell_ls( char* args )
 {
   const DM_DEVICE *pdev;
@@ -203,7 +201,7 @@
   printf( "\n" );
 }
 
-// 'cat' and 'type' handler
+// 'cat' handler
 static void shell_cat( char *args )
 {
   FILE *fp;
@@ -232,7 +230,7 @@
       args = p + 1;
     }      
   else
-      printf( "Usage: cat (or type) <filename1> [<filename2> ...]\n" );
+      printf( "Usage: cat <filename1> [<filename2> ...]\n" );
 }    
 
 // Insert shell commands here
@@ -240,13 +238,13 @@
 {
   { "help", shell_help },
   { "lua", shell_lua },
+#ifdef BUILD_XMODEM
   { "recv", shell_recv },
+#endif
   { "ver", shell_ver },
   { "exit", NULL },
   { "ls", shell_ls },
-  { "dir", shell_ls },
   { "cat", shell_cat },
-  { "type", shell_cat },
   { NULL, NULL }
 };
 
